File: verify/page_settle.js
/* 하루치 정산(SPEC.settleDay) 검증.
 *
 *   python verify/page_settle.py && node verify/page_settle.js
 *
 * settleDay 를 하루씩 이어 돌린 결과가 engine/strategy.py 의 백테스트와
 * 같은 체결을 내는지 대조한다. 페이지가 "어제 이렇게 체결됐어야 한다" 고
 * 제안하는 내용이 백테스트와 어긋나면 안 되기 때문이다.
 */
'use strict';
const fs = require('fs'), path = require('path'), vm = require('vm');
const DIST = path.join(__dirname, '..', 'soxl-daily-orders-source', 'dist');
const EXP = path.join(__dirname, '_settle_expected.json');

if (!fs.existsSync(EXP)) {
  console.error('기대값이 없습니다. 먼저 실행:  python verify/page_settle.py');
  process.exit(1);
}
const sb = { window: {}, console }; sb.globalThis = sb; vm.createContext(sb);
for (const f of ['data.js', 'spec.js']) vm.runInContext(fs.readFileSync(path.join(DIST, f), 'utf8'), sb);
const { SEED, SPEC } = sb.window;
const exp = JSON.parse(fs.readFileSync(EXP, 'utf8'));

console.log('='.repeat(80));
console.log('하루치 정산 vs 백테스트 엔진');
console.log('='.repeat(80));
console.log(`  구간 ${exp.from} ~ ${exp.to}  (${exp.days}일)`);

// 백테스트와 같은 상태에서 출발해 하루씩 정산해 나간다
let state = {
  cash: exp.start.cash,
  lots: exp.start.lots.map((l) => ({ ...l })),
  lastRung: exp.start.lastRung,
  boHold: { ticker: exp.start.boTicker, qty: exp.start.boQty },
};

const norm = (s) => s.split(' ')[0];
let fails = 0, checks = 0, matched = 0;
const got = [];

for (let i = 0; i < exp.steps.length; i++) {
  const step = exp.steps[i];
  const upto = SEED.soxl.findIndex((r) => r[0] === step.date);
  if (upto < 0) { fails++; console.log(`  [실패] ${step.date} 봉을 찾을 수 없습니다`); break; }
  const sIdx = SEED.soxs.findIndex((r) => r[0] === step.date);

  const r = SPEC.settleDay({
    soxl: SEED.soxl.slice(0, upto + 1),
    soxs: SEED.soxs.slice(0, sIdx + 1),
    rsi: step.rsi,
    cash: state.cash, lots: state.lots, lastRung: state.lastRung, boHold: state.boHold,
  });
  if (!r.ok) { fails++; console.log(`  [실패] ${step.date}: ${r.reason}`); break; }

  // 체결 이벤트를 백테스트와 대조
  const mine = r.events.filter((e) => e.sign !== 0)
    .map((e) => `${e.kind}:${e.qty}@${e.price.toFixed(4)}`).sort();
  const want = step.fills.map((f) => `${f.kind}:${f.qty}@${f.price.toFixed(4)}`).sort();
  checks++;
  if (JSON.stringify(mine) !== JSON.stringify(want)) {
    fails++;
    if (fails <= 5) {
      console.log(`  [불일치] ${step.date}`);
      console.log(`      JS  ${mine.join(' | ') || '(없음)'}`);
      console.log(`      PY  ${want.join(' | ') || '(없음)'}`);
    }
  } else matched++;

  got.push({ date: step.date, n: mine.length });
  state = r.next;
}

console.log(`\n  일자 대조 ${checks}일 중 일치 ${matched}일, 불일치 ${fails}일`);
const totalFills = got.reduce((s, x) => s + x.n, 0);
console.log(`  정산된 체결 ${totalFills}건`);
console.log(`\n  최종 상태  현금 $${state.cash.toFixed(2)}  보유 ${state.lots.length}칸  `
  + `마지막칸 ${state.lastRung}  돌파 ${state.boHold.qty ? state.boHold.ticker + ' ' + state.boHold.qty + '주' : '없음'}`);
console.log(`  백테스트   현금 $${exp.end.cash.toFixed(2)}  보유 ${exp.end.lots.length}칸  `
  + `마지막칸 ${exp.end.lastRung}  돌파 ${exp.end.boQty ? exp.end.boTicker + ' ' + exp.end.boQty + '주' : '없음'}`);

const near = (a, b) => Math.abs(a - b) <= 1e-6 * Math.max(1, Math.abs(a), Math.abs(b));
if (!near(state.cash, exp.end.cash)) { fails++; console.log('  [실패] 최종 현금 불일치'); }
if (state.lots.length !== exp.end.lots.length) { fails++; console.log('  [실패] 최종 보유 칸 수 불일치'); }
else state.lots.forEach((l, i) => {
  const w = exp.end.lots[i];
  if (l.no !== w.no || !near(l.px, w.px) || !near(l.qty, w.qty) || l.date !== w.date) {
    fails++; console.log(`  [실패] 보유 칸 ${i}: JS ${JSON.stringify(l)} PY ${JSON.stringify(w)}`);
  }
});
if (state.lastRung !== exp.end.lastRung) { fails++; console.log('  [실패] lastRung 불일치'); }

console.log('\n' + '='.repeat(80));
console.log(fails ? `실패 ${fails}건` : '통과 — 정산 결과가 백테스트와 동일');
process.exit(fails ? 1 : 0);

File: verify/page_settle.py
# -*- coding: utf-8 -*-
"""하루치 정산 대조용 기대값 생성.

engine/strategy.py 의 run() 을 페이지가 실은 구간에서 다시 돌리면서
매일 어떤 체결이 있었는지 기록한다. page_settle.js 가 같은 구간을
SPEC.settleDay() 로 하루씩 이어 돌려 대조한다.
"""
import os, sys, json

H = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.join(H, '..', 'engine'))
import strategy as S
from strategy import load

B = load()
KEEP = 260                       # dist/data.js 와 같은 구간
BW = B[-KEEP:]
WARM = S.MA_LEN + 2              # MA200 이 서는 지점부터
START = WARM

bo = S.Breakout(0.0)
g = S.Grid(0.0)
cash = 100000.0                  # 두 슬리브 합산 현금 하나로 다룬다 (페이지와 동일)
last_rung = 0
steps = []


def lots_json(pos, base):
    return [{'no': p['no'], 'px': p['px'], 'qty': p['q'],
             'date': BW[p['day']]['d'].isoformat(),
             'regime': p['rg']} for p in pos]


start_state = {'cash': cash, 'lots': [], 'lastRung': 0, 'boTicker': None, 'boQty': 0}

for i in range(START, len(BW)):
    bar, prev = BW[i], BW[i - 1]
    fills = []

    # 1) 그 아침의 주문 — 프리장 시점이라 시가를 모른다.
    #    청산대금은 전일 종가로 '추정'해서 수량을 잡는다 (페이지와 동일).
    lots = [{'px': p['px'], 'q': p['q'], 'day_index': p['day'], 'rg': p['rg'], 'no': p['no']}
            for p in g.pos]
    o = S.next_orders(BW[:i], 0.0, cash, bo.sh, lots, regime=bar['rg'],
                      bo_tick=(bo.tick or 'SOXL'), last_rung=last_rung)

    # 2) 장 시작 — 전일 돌파분 MOO 청산 (실제 시가로 대금이 들어온다)
    if bo.sh > 0:
        px = bar['o'] if bo.tick == 'SOXL' else bar['x']['o']
        cash += bo.sh * px * (1 - S.FEE)
        fills.append({'kind': 'bo-exit', 'qty': bo.sh, 'price': px})
        bo.sh = 0.0; bo.tick = None
    gcash = o['리밸런싱']['그리드현금']
    bocash = o['리밸런싱']['돌파현금']

    # 3) 돌파 체결 판정
    b = o['돌파']
    if b and b['수량'] > 0:
        tick = b['종목']
        cb = bar if tick == 'SOXL' else bar['x']
        T, L = b['감시가'], b['지정가']
        fill = None
        if cb and cb['h'] > T:
            if cb['o'] >= T:
                if cb['o'] <= L: fill = cb['o']
                elif cb['l'] <= L: fill = L
            else: fill = L
        if fill is not None:
            q = b['수량']
            cash -= q * fill * (1 + S.FEE)
            bo.sh = q; bo.px = fill; bo.tick = tick
            fills.append({'kind': 'bo-entry', 'qty': q, 'price': fill})

    # 4) 그리드 신규 매수 (프리장 현금 기준)
    ge = o['그리드진입']
    if ge and ge['수량'] > 0 and bar['c'] <= ge['지정가']:
        q = ge['수량']
        cash -= q * bar['c'] * (1 + S.FEE)
        g.pos.append(dict(q=q, px=bar['c'], day=i, tp=S.TP[bar['rg']], rg=bar['rg'], no=ge['칸']))
        last_rung = ge['칸']
        fills.append({'kind': 'grid-buy', 'qty': q, 'price': bar['c']})

    # 5) 그리드 청산
    keep = []
    for p in g.pos:
        if p['day'] == i: keep.append(p); continue
        tp_hit = bar['c'] / p['px'] - 1 >= p['tp']
        forced = i - p['day'] >= S.MAX_DAYS
        if tp_hit or forced:
            cash += p['q'] * bar['c'] * (1 - S.FEE)
            fills.append({'kind': 'grid-tp' if tp_hit else 'grid-moc',
                          'qty': p['q'], 'price': bar['c']})
        else: keep.append(p)
    g.pos = keep
    if not g.pos: last_rung = 0

    if i == START:
        start_state = {'cash': 100000.0, 'lots': [], 'lastRung': 0,
                       'boTicker': None, 'boQty': 0}
    steps.append({'date': bar['d'].isoformat(), 'rsi': bar['rsi'], 'fills': fills})

end = {'cash': cash, 'lots': lots_json(g.pos, BW), 'lastRung': last_rung,
       'boTicker': bo.tick, 'boQty': bo.sh}
payload = {'from': BW[START]['d'].isoformat(), 'to': BW[-1]['d'].isoformat(),
           'days': len(steps), 'start': start_state, 'steps': steps, 'end': end}
p = os.path.join(H, '_settle_expected.json')
json.dump(payload, open(p, 'w', encoding='utf-8'), ensure_ascii=False, indent=1)
nf = sum(len(s['fills']) for s in steps)
print(f'{payload["from"]} ~ {payload["to"]}  {len(steps)}일  체결 {nf}건 -> {p}')
print('이어서:  node verify/page_settle.js')
